text/0000-orderer-pruning-and-snapshot.md

```diff
@@ -0,0 +1,272 @@
+---
+layout: default
+title: Orderer ledger pruning/snapshots
+nav_order: 3
+---
+
+- Feature Name: OSN_pruning_snapshots
+- Start Date: 2021-03-23
+- RFC PR:
+- Fabric Component: orderer, ledger
+- Fabric Issue:
+
+# Summary
+[summary]: #summary
+
+The RFC introduces ledger snapshot feature at orderers, similar to the same feature at peers. Additionally the RFC describes the proposed process of pruning the ledger without restarting the orderer.
+
+# Motivation
+[motivation]: #motivation
+
+Real world use cases show the necessity of controlling the disk space consumed by block ledger, and this RFC is mainly motivated by the need to reduce space consumption at orderer nodes.
+Additionally this RFC presents a better way to join a channel without pulling the huge number of all historical blocks, which also may take significant time and may pose a problem for scaling of an existing Fabric network.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+This RFC introduces a notion of a **ledger snapshot for OSNs** (ordering service nodes). With this notion three new logical operations appear at the orderer REST API and in the CLI tool *osnadmin*:
+- Create a snapshot,
+- Join channel from a snapshot,
+- Prune block ledger of an existing channel.
+
+Ledger snapshot is a set of files containing all the necessary information for OSN to join a channel from a block later than a genesis block (block #0) without loss of any functionality except the ability to deliver an earlier block.
+OSN ledger snapshot functionality is designed with the same principles as the peer ledger snapshot, in particular it is a full snapshot. We do not implement an incremental snapshot capability.
+
+## Guide: Create snapshot
+
+To create a ledger snapshot use the following REST API (block_number must be supplied as a parameter in the request body):
+
+    POST /v1/mgmt/channels/{channelID}/snapshot
+
+ or the following CLI command:
+
+    $ osnadmin snapshot submitrequest -c <channelID> -b <block_number>
+
+A process of snapshot creation will start immediately if the block number is less than current ledger height. If the block number is in future (>= ledger height) then the snapshot creation will be scheduled and the snapshot will be created when the requested block will appear in the ledger. To check the status of the pending snapshot requests use the following REST API:
```

Comment:
I think it's best to simply fail fast in case the block height is higher. We have tools to track a block height, so this can be addressed with machinery instead of making the implementation and API even more complex.

Reply:
This feature (scheduling snapshot in the future) was included only to be consistent with the peer snapshot implementation, where it is the only way of creating snapshots. So the code for snapshot scheduling is already out there.

If general consensus is that the ease of development is more important at this point, we would happily strike out this "scheduling" feature, as we understand that the development and testing will be mostly on us.

I would like to collect a second opinion on this though.

Reply:
Done.
Snapshot scheduling in future will fail

Reply:
Thanks!. As we have agreement on this, following are additional points to consider...

- In my opinion, this API be made a synchronous call as the processing is not expected to take long. It makes the job of the caller easier
- <block_number> can be made optional (or block zero can be treated as a special value), which could be interpreted as the latest committed block

Reply:
Done, here:
27c1fbbadf41bbba9c119d99b6d27fff4ad7ec11